core/src/main/scala/org/http4s/otel4s/middleware/UriRedactor.scala

```diff
@@ -0,0 +1,122 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.otel4s.middleware
+
+import org.http4s.Query
+import org.http4s.Uri
+
+/** Something that redacts URIs to remove sensitive information.
+  *
+  * @note It is highly recommended to use [[UriRedactor.ByParts]] for
+  *       implementations, as it allows easier redaction of individual
+  *       parts of the URI.
+  *
+  * @see [[UriRedactor.ByParts]]
+  */
+trait UriRedactor {
+
+  /** Redacts a URI.
+    *
+    * @note [[UriRedactor.ByParts]] comes with a default implementation of this
+    *       method
+    * @return a redacted URI, or `None` if the entire URI is sensitive
+    */
+  def redact(uri: Uri): Option[Uri]
+}
+
+object UriRedactor {
+
+  /** The string `"REDACTED"`, for use in URI redactors. */
+  final val Redacted = "REDACTED"
+
+  /** A `UriRedactor` that only redacts the userinfo part of the URI's authority. */
+  val OnlyRedactUserInfo: UriRedactor =
+    new ByParts {
+      protected def redactPath(path: Uri.Path): Uri.Path = path
+      protected def redactQuery(query: Query): Query = query
+      protected def redactFragment(
+          fragment: Uri.Fragment
+      ): Option[Uri.Fragment] = Some(fragment)
+    }
+
+  /** A `UriRedactor` that always redacts the entire URI. */
+  val RedactEntirely: UriRedactor = _ => None
+
+  /** A [[`UriRedactor`]] that redacts individual parts of the URI separately. */
+  trait ByParts extends UriRedactor {
+
+    /** Redacts the username and password from a URI's authority. */
+    protected final def redactUserInfo(authority: Uri.Authority): Uri.Authority =
+      authority.userInfo.fold(authority) { info =>
+        authority.copy(userInfo =
+          Some(
+            Uri.UserInfo(
+              username = UriRedactor.Redacted,
+              password = info.password.map(_ => UriRedactor.Redacted),
+            )
+          )
+        )
+      }
+
+    /** @return a redacted authority, or `None` if the entire authority is
+      *         sensitive
+      */
+    protected def redactAuthority(authority: Uri.Authority): Option[Uri.Authority] =
+      Some(redactUserInfo(authority))
+
+    /** @return a redacted path, or `Uri.Path.empty` if the entire path is
+      *         sensitive
+      */
+    protected def redactPath(path: Uri.Path): Uri.Path
+
+    /** @return a redacted query, or `Query.empty` if the entire query is
+      *         sensitive
+      */
+    protected def redactQuery(query: Query): Query
```

Comment:
I'm struggling to remember why these should evaporate to emptiness instead of revealing themselves as `REDACTED`, but I think we had a reason, and it's not hard for anyone who wants to do so.

Reply:
I mean, you can do whichever you want. it's flexible